In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os


In [2]:
#converted project folder path into a string
path=r'C:\Users\manis\Documents\World_covid_data_Analysis'

In [3]:
#Import the “.csv” data set into your notebook
df = pd.read_csv(os.path.join(path, '02 Data', 'original_data', 'main_covid_dataset.csv'))

In [4]:
#Print the first 5 rows of your dataframe
df.head(5)

iso_code continent     location       date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2/24/2020          5.0        5.0   
1      AFG      Asia  Afghanistan  2/25/2020          5.0        0.0   
2      AFG      Asia  Afghanistan  2/26/2020          5.0        0.0   
3      AFG      Asia  Afghanistan  2/27/2020          5.0        0.0   
4      AFG      Asia  Afghanistan  2/28/2020          5.0        0.0   

   total_deaths  new_deaths  icu_patients  hosp_patients  ...  \
0           NaN         NaN           NaN            NaN  ...   
1           NaN         NaN           NaN            NaN  ...   
2           NaN         NaN           NaN            NaN  ...   
3           NaN         NaN           NaN            NaN  ...   
4           NaN         NaN           NaN            NaN  ...   

   population_density  median_age  aged_65_older  aged_70_older  \
0              54.422        18.6          2.581          1.337   
1              54.422        18.6          2.581          1.337   
2              54.422        18.6          2.581          1.337   
3              54.422        18.6          2.581          1.337   
4              54.422        18.6          2.581          1.337   

   gdp_per_capita  cardiovasc_death_rate  diabetes_prevalence  \
0        1803.987                597.029                 9.59   
1        1803.987                597.029                 9.59   
2        1803.987                597.029                 9.59   
3        1803.987                597.029                 9.59   
4        1803.987                597.029                 9.59   

   handwashing_facilities  hospital_beds_per_thousand  life_expectancy  
0                  37.746                         0.5            64.83  
1                  37.746                         0.5            64.83  
2                  37.746                         0.5            64.83  
3                  37.746                         0.5            64.83  
4                  37.746                         0.5            64.83  

[5 rows x 28 columns]

In [5]:
#Total number of Rows and Columns in our dataset
df.shape

(138727, 28)

In [6]:
#Data Wrangling Procedures
#Dropping Columns : Columns which are not relevant for the analysis

df_new=df.drop(columns=['iso_code','icu_patients','hosp_patients','new_tests','total_tests','positive_rate','total_boosters','population_density','median_age','aged_65_older','aged_70_older','gdp_per_capita','cardiovasc_death_rate','diabetes_prevalence','handwashing_facilities','hospital_beds_per_thousand','life_expectancy'])

In [7]:
df_new.shape

(138727, 11)

In [8]:
#Print the name of the columns in new dataset df_new
df_new.columns

Index(['continent', 'location', 'date', 'total_cases', 'new_cases',
       'total_deaths', 'new_deaths', 'total_vaccinations', 'people_vaccinated',
       'people_fully_vaccinated', 'population'],
      dtype='object')

In [9]:
#Renaming the Column
df_new.rename(columns={'total_cases':'cumulative_cases'},inplace=True)

In [10]:
df_new.rename(columns={'total_deaths':'cumulative_deaths'},inplace=True)

In [11]:
# Finding missing values
df_new.isnull().sum()

continent                    8825
location                        0
date                            0
cumulative_cases             7728
new_cases                    7733
cumulative_deaths           19061
new_deaths                  18865
total_vaccinations         102159
people_vaccinated          103780
people_fully_vaccinated    106727
population                    988
dtype: int64

#Except location and date , all the columns have missing entries.For all the columns with number of cases and deaths, for all those missing entries we can record "0" in our dataset, because those days no cases or deaths must have been reported in many different countries. 

#For vaccination columns , highest number of missing records are there because vaccination was approved and started in the later part of the year, so we are going to replace all those missing records with "0" as well

#Continent column missing records shows the lack of information, so here I am going to remove these observations from the dataset, because we dont know to which location that data belongs to.

#Population column missing records belongs to not any particular location, they are categorized as "International" , so will delete all those records as they are not relevant for our countrywise analysis

In [12]:
#Creating a new dataset without all the rows which has continent column as missing value
df_new_clean=df_new[df_new['continent'].isnull()==False]

In [13]:
df_new.shape

(138727, 11)

In [14]:
df_new_clean.shape

(129902, 11)

##### Difference in the shape of Old dataset and new dataset shows that all the 8825 rows with Continent missing values have been removed

##### Imputing missing values of cases, deaths and Vaccination columns with "0"

In [16]:
df_new_clean['cumulative_cases'].fillna(0, inplace=True)

In [17]:
#Checking missing values for column "total_cases"
df_new_clean['cumulative_cases'].isnull().sum()

0

In [18]:
df_new_clean['new_cases'].fillna(0, inplace=True)

In [19]:
df_new_clean['cumulative_deaths'].fillna(0, inplace=True)

In [20]:
df_new_clean['new_deaths'].fillna(0, inplace=True)

In [21]:
df_new_clean['total_vaccinations'].fillna(0, inplace=True)

In [22]:
df_new_clean['people_vaccinated'].fillna(0, inplace=True)

In [23]:
df_new_clean['people_fully_vaccinated'].fillna(0, inplace=True)

In [24]:
# Checking Again for missing values
df_new_clean.isnull().sum()

continent                    0
location                     0
date                         0
cumulative_cases             0
new_cases                    0
cumulative_deaths            0
new_deaths                   0
total_vaccinations           0
people_vaccinated            0
people_fully_vaccinated      0
population                 317
dtype: int64

In [25]:
# Deleting all the rows with missing values in Population column
mod_df = df_new_clean.dropna( how='any', subset=['population'])

In [26]:
# Checking Again for missing values
mod_df.isnull().sum()

continent                  0
location                   0
date                       0
cumulative_cases           0
new_cases                  0
cumulative_deaths          0
new_deaths                 0
total_vaccinations         0
people_vaccinated          0
people_fully_vaccinated    0
population                 0
dtype: int64

#### Now this dataset has no missing values now

In [33]:
mod_df.shape

(129585, 11)

In [27]:
#Duplicates
# Finding Duplicates
#This command will look for full duplicates within df_prods dataframe and create new dataframe df_dups containing only rows that are duplicated
df_dups=mod_df[mod_df.duplicated()]

In [28]:
df_dups

Empty DataFrame
Columns: [continent, location, date, cumulative_cases, new_cases, cumulative_deaths, new_deaths, total_vaccinations, people_vaccinated, people_fully_vaccinated, population]
Index: []

In [29]:
# Looking for mixed datatypes
mod_df.dtypes

continent                   object
location                    object
date                        object
cumulative_cases           float64
new_cases                  float64
cumulative_deaths          float64
new_deaths                 float64
total_vaccinations         float64
people_vaccinated          float64
people_fully_vaccinated    float64
population                 float64
dtype: object

In [30]:
for col in mod_df.columns.tolist():
  weird = (mod_df[[col]].applymap(type) != mod_df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (mod_df[weird]) > 0:
    print (col)

#### This dataset has no mixed datatypes

In [31]:
mod_df.describe()

cumulative_cases      new_cases  cumulative_deaths     new_deaths  \
count      1.295850e+05  129585.000000      129585.000000  129585.000000   
mean       5.276690e+05    2062.197122       11711.378254      40.608543   
std        2.618355e+06   10875.280753       50178.067344     189.245485   
min        0.000000e+00  -74347.000000           0.000000   -1918.000000   
25%        1.300000e+03       0.000000          20.000000       0.000000   
50%        1.744600e+04      60.000000         297.000000       1.000000   
75%        1.929940e+05     697.000000        3501.000000      10.000000   
max        4.953895e+07  414188.000000      793228.000000    7374.000000   

       total_vaccinations  people_vaccinated  people_fully_vaccinated  \
count        1.295850e+05       1.295850e+05             1.295850e+05   
mean         7.845753e+06       3.034485e+06             1.967066e+06   
std          8.342807e+07       2.527370e+07             1.670698e+07   
min          0.000000e+00       0.000000e+00             0.000000e+00   
25%          0.000000e+00       0.000000e+00             0.000000e+00   
50%          0.000000e+00       0.000000e+00             0.000000e+00   
75%          0.000000e+00       0.000000e+00             0.000000e+00   
max          2.574931e+09       1.225000e+09             1.125000e+09   

         population  
count  1.295850e+05  
mean   4.039324e+07  
std    1.500292e+08  
min    4.700000e+01  
25%    1.748295e+06  
50%    8.478242e+06  
75%    2.967492e+07  
max    1.444216e+09

In [32]:
# Exported the df_new_clean dataset in prepared_data folder as CSV file MainCoviddata_Checked.csv
df_new_clean.to_csv(os.path.join(path, '02 Data','prepared_data', 'maincoviddataset_checked.csv'))